### PR TITLE
Switch to local repo for pytests. Fix broken fedora builds

### LINKED ISSUE #100: fedora dist tests failing
@cbsiddharth can you please take a look at this?

### PR DESCRIPTION
**Pull request by sidcha**
- Create a local repo and use only that for all tests.
- Adds some more tests.
- Fix broken fedora builds. Fixes #100.

### CODE REVIEW

**Comment by sidcha** [comment]
@keerthanakalyan can you please prioritize and review this?

File: pytests/conftest.py [hunk 1]
```diff
@@ -94,36 +118,55 @@ def wget(self, url, out, enforce_https=True, fingerprint=None):
...
         return ret
 
-def backup_config_files():
-    # Backup /etc/yum.repos.d/* and /etc/tdnf/tdnf.conf
-    pass
-
-def restore_config_files():
-    # Restore /etc/yum.repos.d/* and /etc/tdnf/tdnf.conf
-    pass
+def backup_files(conf_dir):
```
Comment (keerthanakalyan) [inline pytests/conftest.py]: I do not find this function's usage in here. Is this used somewhere else?

File: pytests/conftest.py [hunk 2]
```diff
@@ -94,36 +118,55 @@ def wget(self, url, out, enforce_https=True, fingerprint=None):
...
+    for file in os.listdir(conf_dir):
+        if file.endswith('.bak'):
+            continue
+        src  = os.path.join(conf_dir, file)
+        dest = os.path.join(conf_dir, os.path.basename(file) + '.bak')
+        shutil.copyfile(src, dest)
+
+def restore_files(conf_dir):
```
Comment (keerthanakalyan) [inline pytests/conftest.py]: I do not find this function's usage in here. Is this used somewhere else?

File: pytests/tests/test_list.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    helper_test_list_sub_cmd(utils, 'recent')
+
+def test_list_upgrades(utils):
+    mpkg = utils.config["mulversion_pkgname"]
+    mpkg_version = utils.config["mulversion_lower"]
+    spkg = utils.config["sglversion_pkgname"]
+
+    ret = utils.run([ 'tdnf', 'list', 'upgrades' ])
+
+    # TODO Fix this; see issue #94
```
Comment (keerthanakalyan) [inline pytests/tests/test_list.py:51]: Please make a note in the issue #94 to fix this corresponding tests as well.

**Review by keerthanakalyan — COMMENTED** [review]
LGTM.

File: pytests/conftest.py (hunk 1 above)
Comment (sidcha) [reply pytests/conftest.py]: It was an unused method. Removed.

File: pytests/tests/test_list.py (hunk 3 above)
Comment (sidcha) [reply pytests/tests/test_list.py:51]: This comment you have here is reference enough (it shows up on that issue lising). Marking this as resolved.

File: pytests/conftest.py (hunk 2 above)
Comment (sidcha) [reply pytests/conftest.py]: It was an unused method. Removed.

**Review by keerthanakalyan — APPROVED** [review]
LGTM

File: ci/Dockerfile.fedora [hunk 4]
```diff
@@ -4,7 +4,7 @@ MAINTAINER csiddharth@vmware.com
 
 
 RUN dnf -q -y upgrade
-RUN dnf -q -y install gcc make cmake libcurl-devel rpm-devel libsolv-devel \
+RUN dnf -q -y install gcc make cmake libcurl-devel rpm-build libsolv-devel \
```
Comment (ppadmavilasom) [inline ci/Dockerfile.fedora]: please keep rpm-devel even if rpm-build brings rpm-devel

File: pytests/conftest.py [hunk 5]
```diff
@@ -94,37 +118,40 @@ def wget(self, url, out, enforce_https=True, fingerprint=None):
...
-
         return True, None
 
     def run(self, cmd):
-        if cmd[0] is 'tdnf' and self.config['build_dir']:
-            cmd[0] = os.path.join(self.config['build_dir'], 'bin/tdnf')
+        if cmd[0] is 'tdnf':
+            if 'build_dir' in self.config:
+                cmd[0] = os.path.join(self.config['build_dir'], 'bin/tdnf')
+            if cmd[1] is not '--config':
```
Comment (ppadmavilasom) [inline pytests/conftest.py:127]: why dont we add -c for all commands without this check? do we have cases where we do not use the test config?

File: pytests/repo/setup-repo.sh [hunk 6]
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+#
+# Copyright (C) 2015 VMware, Inc. All Rights Reserved.
```
Comment (ppadmavilasom) [inline pytests/repo/setup-repo.sh]: change copyright year

File: pytests/tests/test_config.py [hunk 7]
```diff
@@ -12,32 +12,30 @@
...
+    utils.run([ 'sed', '-i', '$ a repodir=/tmp/myrepo', '/tmp/myrepo/mytdnf.conf' ])
+    utils.run([ 'cp', tdnf_repo, '/tmp/myrepo/photon.repo' ])
+    utils.run([ 'sed', '-i', 's/enabled=0/enabled=1/g', '/tmp/myrepo/photon.repo' ])
     yield
     teardown_test(utils)
 
 def teardown_test(utils):
-    utils.run([ 'cp', '/etc/yum.repos.d/photon.repo.bak', '/etc/yum.repos.d/photon.repo' ])
-    utils.run([ 'rm', '/etc/tdnf/mytdnf.conf' ])
     utils.run([ 'rm', '-rf', '/etc/myrepo' ])
```
Comment (ppadmavilasom) [inline pytests/tests/test_config.py]: setup_test moved all this to /tmp/myrepo so this is not required

File: pytests/tests/test_erase.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
+#
+# Copyright (C) 2019 VMware, Inc. All Rights Reserved.
```
Comment (ppadmavilasom) [inline pytests/tests/test_erase.py]: 2020

File: pytests/tests/test_install.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
+#
+# Copyright (C) 2019 VMware, Inc. All Rights Reserved.
```
Comment (ppadmavilasom) [inline pytests/tests/test_install.py]: 2020

File: pytests/conftest.py (hunk 5 above)
Comment (sidcha) [reply pytests/conftest.py:127]: When we are testing `--config` from tests, we must pass newly created config file (or dummy ones) to validate that the passed config is being considered. In those cases, this method must not pass any config file.

File: pytests/repo/setup-repo.sh (hunk 6 above)
Comment (sidcha) [reply pytests/repo/setup-repo.sh]: Oops. Fixed.

File: ci/Dockerfile.fedora (hunk 4 above)
Comment (sidcha) [reply ci/Dockerfile.fedora]: done.

File: pytests/tests/test_config.py (hunk 7 above)
Comment (sidcha) [reply pytests/tests/test_config.py]: Must have been `/tmp/myrepo`. Fixed

File: pytests/tests/test_erase.py (hunk 8 above)
Comment (sidcha) [reply pytests/tests/test_erase.py]: these commits were authored in dec 2019. The commit dates also reflect this. Do you still want me to put them to 2020?

File: pytests/tests/test_install.py (hunk 9 above)
Comment (sidcha) [reply pytests/tests/test_install.py]: these commits were authored in dec 2019. The commit dates also reflect this. Do you still want me to put them to 2020?

File: pytests/tests/test_erase.py (hunk 8 above)
Comment (ppadmavilasom) [reply pytests/tests/test_erase.py]: please do 2019 - 2020 then

Comment (sidcha) [reply pytests/tests/test_erase.py]: Done.

File: pytests/tests/test_install.py (hunk 9 above)
Comment (sidcha) [reply pytests/tests/test_install.py]: Copyright year set to 2019 - 2020.

**Review by ppadmavilasom — APPROVED** [review]
Thanks much for the fixes. Appreciate it.

**Comment by sidcha** [comment]
> Thanks much for the fixes. Appreciate it.

@ppadmavilasom, the work for this PR started as soon as the first PR was submitted. Somewhere along the way got into other activities. I wish I had submitted this earlier.

One particular upside with this change that I'm happy about is that tests run significantly faster.

**Pull request merged**